### PR TITLE
Send auth via AMQP and Remove proto from knotd

### LINKED ISSUE #12: Authenticate devices
- [ ] List devices on start
- [ ] Send 'authenticate' command to message queue
- [ ] Implement 'authenticate' on connector

### LINKED ISSUE #13: Remove leftover code

### PR DESCRIPTION
**Pull request by vitbaq**
Closes #12 
Closes #13

### CODE REVIEW

File: src/msg.c [hunk 1]
```diff
@@ -1510,6 +1510,8 @@ static bool on_cloud_receive(const struct cloud_msg *msg, void *user_data)
 		return handle_device_added(session, msg->device_id, msg->token);
 	case UNREGISTER_MSG:
 		return handle_device_removed(msg->device_id);
+	case AUTH_MSG:
+		return NULL;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Done

File: src/cloud.c [hunk 2]
```diff
@@ -147,6 +150,13 @@ static struct cloud_msg *create_msg(const char *routing_key, json_object *jso)
 			goto err;
 		}
 
+		break;
+	case AUTH_MSG:
+		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
+		msg->auth = NULL;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Done

File: src/msg.c [hunk 3]
```diff
@@ -557,8 +504,8 @@ static int8_t msg_auth(struct session *session,
 {
 	char uuid[KNOT_PROTOCOL_UUID_LEN + 1];
 	char token[KNOT_PROTOCOL_TOKEN_LEN + 1];
+	char id[KNOT_ID_LEN];
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Done

File: src/cloud.c [hunk 4]
```diff
@@ -311,8 +312,31 @@ int cloud_unregister_device(const char *id)
...
+	int result;
+
+	jobj_device = NULL;
+	json_str = json_object_to_json_string(jobj_device);
+
+	queue_cloud = amqp_declare_new_queue(AMQP_QUEUE_CLOUD);
+	if (queue_cloud.bytes == NULL) {
+		hal_log_error("Error on declare a new queue.\n");
+		return -1;
+	}
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Done

File: src/msg.c [hunk 5]
```diff
@@ -1370,23 +1253,11 @@ static bool handle_cloud_msg_list(struct l_queue *devices)
 static void start_timeout(struct l_timeout *timeout, void *user_data)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Done

File: README [hunk 6]
```diff
@@ -16,16 +16,12 @@ ell >= 0.4
...
 libssl-dev
 valgrind (optional)
 
 How to install dependencies:
 	$sudo apt-get install automake libtool libssl-dev valgrind
-	To install libwebsockets first you need to install cmake
-		https://cmake.org/install/
-	After install cmake, install libwebsocket v2.4.2
-		https://github.com/warmcat/libwebsockets/releases/tag/v2.4.2
+
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply README]: Done

File: src/manager.c [hunk 7]
```diff
@@ -43,212 +43,11 @@
 #include "storage.h"
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/manager.c]: Done

File: src/settings.c [hunk 8]
```diff
@@ -127,18 +112,13 @@ struct settings *settings_load(int argc, char *argv[])
 {
 	struct settings *settings;
 	struct stat buf;
-	unsigned int port = 0;
-	char *host = NULL;
 	char *uuid = NULL;
 	char *token = NULL;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/settings.c]: Done

File: src/parser.c [hunk 9]
```diff
@@ -781,3 +781,16 @@ const char *parser_get_key_str_from_json_obj(json_object *jso, const char *key)
...
+{
+	json_object *jobjkey;
+
+	if (!json_object_object_get_ex(jso, key, &jobjkey))
+		return false;
+
+	if (json_object_get_type(jobjkey) != json_type_boolean)
+		return false;
+
+	return json_object_get_boolean(jobjkey);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/parser.c]: Done

File: src/msg.c [hunk 10]
```diff
@@ -1317,6 +1317,14 @@ static bool handle_device_removed(const char *device_id)
 	return false;
 }
 
+static bool handle_device_auth(struct session *session, int auth)
+{
+	if (auth)
+		session->trusted = auth;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Done

File: src/cloud.c [hunk 11]
```diff
@@ -298,6 +298,23 @@ int cloud_unregister_device(const char *id)
...
+ * @id: device id
+ * @token: device token
+ *
+ * Requests cloud to auth a device.
+ * The confirmation that the cloud received the message comes from a callback
+ * set in function cloud_set_read_handlers.
+ *
+ * Returns: 0 if successful and a KNoT error otherwise.
+ */
+int cloud_auth_device(const char *id, const char *token)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Done

File: src/msg.c [hunk 12]
```diff
@@ -370,59 +370,6 @@ static void downstream_callback(struct l_timeout *timeout, void *user_data)
...
-	/* FIXME: manage link overload or not connected */
-	session = l_queue_find(session_list, session_node_fd_cmp, user_data);
-	if (!session)
-		return false;
-
-	if (strcmp("config", name) == 0) {
-		if (session->config && strcmp(session->config, value) == 0)
-			goto done;
-
-		list = parser_config_to_list(value);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Outdated

File: src/msg.c [hunk 13]
```diff
@@ -370,59 +370,6 @@ static void downstream_callback(struct l_timeout *timeout, void *user_data)
...
-	struct knot_device *device;
-	char id[KNOT_ID_LEN];
-
-	/* FIXME: manage link overload or not connected */
-	session = l_queue_find(session_list, session_node_fd_cmp, user_data);
-	if (!session)
-		return false;
-
-	if (strcmp("config", name) == 0) {
-		if (session->config && strcmp(session->config, value) == 0)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Done

File: src/cloud.c [hunk 14]
```diff
@@ -300,6 +300,21 @@ int cloud_unregister_device(const char *id)
 	return 0;
 }
 
+/**
+ * cloud_auth_device:
+ *
+ * Requests cloud to auth a device.
+ * The confirmation that the cloud received the message comes from a callback
+ * set in function cloud_set_read_handlers.
```
Comment (ramonhpr) [inline src/cloud.c]: *cloud_set_read_handler*

File: src/msg.c [hunk 15]
```diff
@@ -1517,9 +1377,8 @@ int msg_start(struct settings *settings)
 	}
 
 	device_id_list = l_queue_new();
-	start_to =  l_timeout_create_ms(1, start_timeout, settings, NULL);
 
-	return (start_to ? 0 : -ENOMEM);
+	cloud_list_devices();
```
Comment (ramonhpr) [inline src/msg.c]: Since we don't have a l_timeout to keep retring we should verify the return of this function and try again if had some error

File: src/settings.c [hunk 16]
```diff
@@ -108,56 +101,17 @@ static int parse_args(int argc, char *argv[], struct settings *settings)
 struct settings *settings_load(int argc, char *argv[])
 {
 	struct settings *settings;
-	struct stat buf;
```
Comment (ramonhpr) [inline src/settings.c]: Since we don't use this struct anymore we can remove header `sys/stat.h`

File: src/msg.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: PR related: https://github.com/CESARBR/knot-service-source/pull/37

File: src/msg.c (hunk 15 above)
Comment (ramonhpr) [reply src/msg.c]: :eyes: You should also return 0 if success otherwise it will stop all the services that you started

**Comment by ramonhpr** [comment]
Missing mock implementation

File: src/msg.c [hunk 17]
```diff
@@ -578,7 +518,7 @@ static int8_t msg_auth(struct session *session,
 	/* Set UUID & token: Used at property_changed */
 	session->uuid = l_strdup(uuid);
 	session->token = l_strdup(token);
-	proto_sock = l_io_get_fd(session->proto_channel);
+
 	/* Set Id */
 	mydevice = l_queue_find(device_id_list, device_uuid_cmp, session->uuid);
```
Comment (ramonhpr) [inline src/msg.c]: You must verify if mydevice is NULL and avoid access it

File: src/msg.c [hunk 18]
```diff
@@ -587,8 +527,7 @@ static int8_t msg_auth(struct session *session,
 	hal_log_info("[session %p] Authenticating UUID: %s, TOKEN: %s",
 		     session, uuid, token);
 
-	result = proto_signin(proto_sock, uuid, token, property_changed,
-			      L_INT_TO_PTR(session->node_fd));
+	result = cloud_auth_device(mydevice->id, token);
```
Comment (ramonhpr) [inline src/msg.c]: It will lead to a segmentation fault if mydevice is not found

Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c (hunk 17 above)
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/settings.c (hunk 16 above)
Comment (vitbaq) [reply src/settings.c]: Ok

File: src/msg.c (hunk 15 above)
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/cloud.c (hunk 14 above)
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/msg.c [hunk 19]
```diff
@@ -886,7 +886,11 @@ static ssize_t msg_process(struct session *session,
 			l_timeout_create_ms(512,
 					    downstream_callback,
 					    session, NULL);
-		break;
+		/*
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/msg.c]: ... to handle **authentication** messages **received**...

(please unify in all commit: authenticated/authentication messages?)

File: src/cloud.c [hunk 20]
```diff
@@ -153,7 +153,8 @@ static struct cloud_msg *create_msg(const char *routing_key, json_object *jso)
 		break;
 	case AUTH_MSG:
 		msg->device_id = parser_get_key_str_from_json_obj(jso, "id");
-		msg->auth = false;
+		msg->auth = parser_get_key_bool_from_json_obj(jso,
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: a new **parser** function...

File: src/msg.c [hunk 21]
```diff
@@ -47,7 +47,6 @@
 #include "node.h"
 #include "device.h"
 #include "proxy.h"
-#include "proto.h"
 #include "cloud.h"
```
Comment (ramonhpr) [inline src/msg.c]: About commit message 597af95989f9cd0025a5b925cfc2a6c9adc44236:
Remove proto_start function from msg source file

File: Makefile.am [hunk 22]
```diff
@@ -13,9 +13,6 @@ noinst_PROGRAMS = tools/ktool unit/ktest unit/inettest
 dbusdir = @DBUS_CONFDIR@/dbus-1/system.d
 dbus_DATA = src/knot.conf
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply Makefile.am]: This patch **removes** ...

File: src/cloud.c [hunk 23]
```diff
@@ -53,6 +53,7 @@
 #define AMQP_EVENT_DATA_REQUEST "data.request"
 #define AMQP_EVENT_DEVICE_REGISTERED "device.registered"
 #define AMQP_EVENT_DEVICE_UNREGISTERED "device.unregistered"
+#define AMQP_EVENT_DEVICE_AUTH "device.auth"
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/msg.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c [hunk 24]
```diff
@@ -61,7 +61,6 @@ struct session {
 	bool trusted;			/* Authenticated */
 	struct node_ops *node_ops;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/manager.c [hunk 25]
```diff
@@ -40,215 +40,13 @@
 #include "settings.h"
 #include "msg.h"
 #include "dbus.h"
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/manager.c]: Ok

File: src/proto.c [hunk 26]
```diff
@@ -46,12 +46,6 @@
 extern struct proto_ops proto_ws;
 extern struct proto_ops proto_socketio;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/proto.c]: Ok

File: Makefile.am (hunk 22 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply Makefile.am]: Ok

File: src/msg.c (hunk 21 above)
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/cloud.c [hunk 27]
```diff
@@ -300,6 +300,21 @@ int cloud_unregister_device(const char *id)
 	return 0;
 }
 
+/**
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/msg.c [hunk 28]
```diff
@@ -1317,6 +1321,33 @@ static bool handle_device_removed(const char *device_id)
 	return false;
 }
 
+static bool handle_device_auth(struct session *session, int auth)
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c [hunk 29]
```diff
@@ -383,30 +380,7 @@ static bool property_changed(const char *name,
...
-			hal_log_error("[session %p] config: parse error!",
-				      session);
-			goto done;
-		}
-
-		if (parser_config_is_valid(list) != 0) {
-			hal_log_error("[session %p] config: invalid format!",
-				      session);
-			l_queue_destroy(list, l_free);
-			goto done;
```
Comment (ramonhpr) [inline src/msg.c]: Remove `done` label to avoid the following warning:
```
src/msg.c: In function ‘property_changed’:
src/msg.c:396:1: error: label ‘done’ defined but not used [-Werror=unused-label]
 done:
 ^~~~
cc1: all warnings being treated as errors
Makefile:926: recipe for target 'src/src_knotd-msg.o' failed
make[1]: *** [src/src_knotd-msg.o] Error 1
Makefile:496: recipe for target 'all' failed
make: *** [all] Error 2
```

Comment (vitbaq) [reply src/msg.c]: Ok

File: src/proto.c [hunk 30]
```diff
@@ -73,21 +67,6 @@ static inline bool is_token_valid(const char *token)
...
-	for (i = 0; proto_ops[i]; i++) {
-		if (strcmp(protocol_name, proto_ops[i]->name) != 0)
-			continue;
-
-		selected_protocol = proto_ops[i];
-	}
-
-	return selected_protocol;
-}
-
```
Comment (ramonhpr) [inline src/proto.c]: This change doesn't seem to belong to this commit.

File: src/proto.c [hunk 31]
```diff
@@ -97,13 +76,7 @@ int proto_start(const struct settings *settings)
 	 * TODO: later support dynamic protocol selection.
 	 */
 
-	proto = get_proto_ops(settings->proto);
-	if (proto == NULL)
-		return -EINVAL;
-
-	hal_log_info("proto_ops: %s", proto->name);
-
-	return proto->probe(settings->host, settings->port);
```
Comment (ramonhpr) [inline src/proto.c]: This change doesn't seem to belong to this commit.

File: Makefile.am (hunk 22 above)
[the remark this answers is not in the document]
Comment (ramonhpr) [reply Makefile.am]: Update commit title too

File: src/msg.c [hunk 32]
```diff
@@ -546,6 +546,9 @@ static int8_t msg_auth(struct session *session,
 		return result;
 	}
 
+	l_queue_foreach(mydevice->schema, schema_dup_foreach,
```
Comment (ramonhpr) [inline src/msg.c]: This code only handle devices that have schema in device_id_list.

Comment (ramonhpr) [reply src/msg.c]: After we receive schema we should duplicate from session->schema_list to device_id_list too.
Because new devices registered are added to device_id_list and don't have schema

File: src/cloud.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/manager.c (hunk 25 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/manager.c]: Ok

File: src/proto.c (hunk 26 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/proto.c]: Ok

File: Makefile.am (hunk 22 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply Makefile.am]: Ok

File: src/msg.c (hunk 32 above)
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/manager.c [hunk 33]
```diff
@@ -40,215 +40,13 @@
 #include "settings.h"
 #include "msg.h"
```
Comment (ramonhpr) [inline src/manager.c]: About commit msg 071f724d0b28b6827cb5c326d01a173700697640:
... by **the** AMQP ...

File: src/proto.c (hunk 26 above)
Comment (ramonhpr) [inline src/proto.c]: About commit message 9c74b77eab5004f097bf63ee60f60da7325188bb:
Theses options **were** ...

File: src/cloud.c [hunk 34]
```diff
@@ -300,6 +300,21 @@ int cloud_unregister_device(const char *id)
...
+ * cloud_auth_device:
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/cloud.c [hunk 35]
```diff
@@ -61,6 +61,7 @@
 #define AMQP_CMD_DATA_PUBLISH "data.publish"
 #define AMQP_CMD_DEVICE_REGISTER "device.register"
 #define AMQP_CMD_DEVICE_UNREGISTER "device.unregister"
+#define AMQP_CMD_DEVICE_AUTH "device.cmd.auth"
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/msg.c [hunk 36]
```diff
@@ -70,7 +70,6 @@ struct session {
 	char *token;			/* Device token */
 	struct l_queue *schema_list;	/* Schema accepted by cloud */
 	struct l_queue *config_list;	/* knot_config accepted from cloud */
-	char *config;			/* Current config */
 	struct l_timeout *downstream_to; /* Active when there is data to send */
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c [hunk 37]
```diff
@@ -546,6 +546,9 @@ static int8_t msg_auth(struct session *session,
 		return result;
 	}
 
+	l_queue_foreach(mydevice->schema, schema_dup_foreach,
+			session->schema_list);
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/msg.c [hunk 38]
```diff
@@ -1420,21 +1420,13 @@ static void start_timeout(struct l_timeout *timeout, void *user_data)
 	if (sock < 0)
 		goto connect_fail;
```
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/msg.c]: Ok

File: src/manager.c (hunk 33 above)
Comment (vitbaq) [reply src/manager.c]: Ok

File: src/proto.c (hunk 26 above)
Comment (vitbaq) [reply src/proto.c]: Ok

File: src/cloud.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (vitbaq) [reply src/cloud.c]: Ok

File: src/proto.c (hunk 31 above)
Comment (vitbaq) [reply src/proto.c]: It is related. I remove settings->proto/hoss/port from the settings source file.

File: src/proto.c (hunk 30 above)
Comment (vitbaq) [reply src/proto.c]: It is related. I remove settings->proto/hoss/port from the settings source file.
